File: notification-service/src/main/java/ru/javavlsu/kb/notificationservice/kafka/KafkaConsumer.java
package ru.javavlsu.kb.notificationservice.kafka;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.stereotype.Component;
import ru.javavlsu.kb.notificationservice.service.EmailService;
import ru.javavlsu.kb.notificationservice.service.NotificationService;

@Component
public class KafkaConsumer {

    private final ObjectMapper objectMapper;
    private final EmailService emailService;
    private final NotificationService notificationService;

    @Autowired
    public KafkaConsumer(ObjectMapper objectMapper, EmailService emailService, NotificationService notificationService) {
        this.objectMapper = objectMapper;
        this.emailService = emailService;
        this.notificationService = notificationService;
    }

    @KafkaListener(topics = "esap.users.mail", groupId = "notification-service")
    public void consumeUserData(@Payload String message,
                                @Header("type") String type) throws JsonProcessingException {
        MessageType messageType = MessageType.valueOf(type);
        if (messageType == MessageType.COMMAND) {
            JsonNode userData = objectMapper.readTree(message);

            String emailSubject = "Добро пожаловать в нашу клинику";
            String toAddress = userData.get("email").asText();
            String emailMessage = "Уважаемый " + userData.get("firstName").asText()
                    + "!\nВысылаем вам данные для личного кабинета"
                    + "\nЛогин: " + userData.get("login").asText() + "\nПароль: " + userData.get("password").asText()
                    + "\n\nС уважением, поликлиника \"" + userData.get("clinic").get("name").asText() + "\".";

            emailService.sendEmail(toAddress, emailSubject, emailMessage);
        }
    }

    @KafkaListener(topics = "esap.users.notifications.mobile", groupId = "notification-service")
    public void consumeUserNotification(@Payload String message,
                                        @Header("type") String type) throws JsonProcessingException {
        MessageType messageType = MessageType.valueOf(type);
        if (messageType == MessageType.COMMAND) {
            JsonNode notificationData = objectMapper.readTree(message);

            String to = notificationData.get("to").asText();
            String title = notificationData.get("title").asText();
            String body = notificationData.get("body").asText();

            notificationService.sendNotificationByToken(to, title, body);
        }
    }
}

File: notification-service/src/main/kotlin/ru/javavlsu/kb/notificationservice/kafka/KafkaConsumer.kt
package ru.javavlsu.kb.notificationservice.kafka

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.stereotype.Component
import ru.javavlsu.kb.notificationservice.dto.NotificationEvent
import ru.javavlsu.kb.notificationservice.dto.TokenRegistrationEvent
import ru.javavlsu.kb.notificationservice.dto.PatientCreatedEvent
import ru.javavlsu.kb.notificationservice.service.email.EmailService
import ru.javavlsu.kb.notificationservice.service.NotificationService
import ru.javavlsu.kb.notificationservice.service.UserDeviceTokenService
import ru.javavlsu.kb.notificationservice.service.email.EmailBuilder

@Component
class KafkaConsumer @Autowired constructor(
    val objectMapper: ObjectMapper,
    val emailService: EmailService,
    val notificationService: NotificationService,
    val userDeviceTokenService: UserDeviceTokenService,
    val emailBuilder: EmailBuilder
) {

    @Throws(JsonProcessingException::class)
    @KafkaListener(topics = ["\${notification.kafka.topic.welcome-email}"], groupId = "notification-service")
    fun consumePatientCreatedEvent(@Payload message: String) {
        val patientCreatedEvent = objectMapper.readValue(message,
            PatientCreatedEvent::class.java)
        val emailContent = emailBuilder.build(patientCreatedEvent)
        emailService.sendEmail(emailContent)
    }

    @Throws(JsonProcessingException::class)
    @KafkaListener(topics = ["\${notification.kafka.topic.token-registration}"], groupId = "notification-service")
    fun consumeTokenRegistrationEvent(@Payload message: String) {
        val tokenRegistrationEvent = objectMapper.readValue(message,
            TokenRegistrationEvent::class.java)
        userDeviceTokenService.saveToken(tokenRegistrationEvent)
    }

    @Throws(JsonProcessingException::class)
    @KafkaListener(topics = ["\${notification.kafka.topic.push-notification}"], groupId = "notification-service")
    fun consumeNotificationEvent(@Payload message: String) {
        val notificationEvent = objectMapper.readValue(message,
            NotificationEvent::class.java)
        notificationService.sendNotificationToUser(notificationEvent)
    }
}